query-engine/query-engine-c-abi/src/migrations.rs

```diff
@@ -0,0 +1,191 @@
+use indoc::indoc;
+use rusqlite::Connection;
+use std::{
+    fs::{read_dir, DirEntry},
+    path::{Path, PathBuf},
+};
+use user_facing_errors::{Error, UnknownError};
+
+pub type Timestamp = chrono::DateTime<chrono::Utc>;
+
+// TODO there is a bunch of casting that is present, however it is not the most correct way
+// but since this is an out of tree branch, I do not want to change the common libraries yet
+
+#[derive(Debug)]
+pub struct MigrationDirectory {
+    path: PathBuf,
+}
+
+impl MigrationDirectory {
+    /// Initialize a MigrationDirectory at the provided path. This will not
+    /// validate that the path is valid and exists.
+    pub fn new(path: PathBuf) -> MigrationDirectory {
+        MigrationDirectory { path }
+    }
+
+    /// The `{timestamp}_{name}` formatted migration name.
+    pub fn migration_name(&self) -> &str {
+        self.path
+            .file_name()
+            .expect("MigrationDirectory::migration_id")
+            .to_str()
+            .expect("Migration directory name is not valid UTF-8.")
+    }
+
+    /// Read the migration script to a string.
+    pub fn read_migration_script(&self) -> Result<String, Error> {
+        let path = self.path.join("migration.sql");
+        std::fs::read_to_string(path).map_err(|ioerr| UnknownError::new(&ioerr).into())
+    }
+}
+
+impl From<DirEntry> for MigrationDirectory {
+    fn from(entry: DirEntry) -> MigrationDirectory {
+        MigrationDirectory { path: entry.path() }
+    }
+}
+
+/// An applied migration, as returned by list_migrations.
+#[derive(Debug, Clone)]
+pub struct MigrationRecord {
+    /// A unique, randomly generated identifier.
+    pub id: String,
+    /// The timestamp at which the migration completed *successfully*.
+    pub finished_at: Option<Timestamp>,
+    /// The name of the migration, i.e. the name of migration directory
+    /// containing the migration script.
+    pub migration_name: String,
+    /// The time the migration started being applied.
+    pub started_at: Timestamp,
+}
+
+pub fn list_migration_dir(migrations_directory_path: &Path) -> Result<Vec<MigrationDirectory>, Error> {
+    let mut entries: Vec<MigrationDirectory> = Vec::new();
+
+    let read_dir_entries = match read_dir(migrations_directory_path) {
+        Ok(read_dir_entries) => read_dir_entries,
+        // Err(err) if matches!(err.kind(), std::io::ErrorKind::NotFound) => return Ok(entries),
+        Err(err) => return Err(UnknownError::new(&err).into()),
+    };
+
+    for entry in read_dir_entries {
+        let entry = entry.map_err(|err| UnknownError::new(&err))?;
+
+        if entry.file_type().map_err(|err| UnknownError::new(&err))?.is_dir() {
+            entries.push(entry.into());
+        }
+    }
+
+    entries.sort_by(|a, b| a.migration_name().cmp(b.migration_name()));
+
+    Ok(entries)
+}
+
+// pub fn detect_failed_migrations(migrations_from_database: &[MigrationRecord]) -> Result<(), user_facing_errors::Error> {
```

Comment:
Should this comment go away? Is the function important?

Reply:
Not yet, as I mentioned in one of the comments above, error handling needs to be handled differently when running on multiple devices at once.

Reply:
I understand that but what's the plan wrt. this function specifically? It's fully commented out. Do you plan on having a follow-up PR? Is that for us to figure out later? Either way, if you want to keep it fully commented out, a TODO comment explaining why it's commented out would be useful

Reply:
yeah, I still don't know how it is meant to behave, the plan for now was just to keep it simple, still need to discuss with soren probably about this.

Reply:
Alright, just talked it over to Soren, the consensus is: if a migration fails, you are done (as we cannot guarantee synchronicity between the schema, generated client, and db state). So I will re-add this function and it will always throw whenever a failed migration is detected.